ImGui-Driven Extraction Flow and Progress Reporting

Malkierian opened the pull request:
This sets up the initialization process to allow for producing an ImGui window before ROM archives are made, so that ImGui can be used to prompt users along the O2R creation process instead of using the OS prompts, which aren't always usable even on some Linux devices, let alone other 2nd-party port platforms, as well as graphically report extraction progress.

https://github.com/user-attachments/assets/b935bffd-e8c7-4be6-9e4e-646e2a942149

 Requires https://github.com/HarbourMasters/ZAPDTR/pull/33 and https://github.com/Kenix3/libultraship/pull/949

TODO: 
 - [x] Convert the O2R update process to this new flow.
 - [x] Figure out why the titlebar close button isn't closing SoH while in this flow.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/5107875127.zip)
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/5107884143.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/5114683330.zip)
<!--- section:artifacts:end -->

Malkierian committed:
Initialize Fast3dWindow before ROM archive checks/extraction.

Malkierian committed:
Start isolation of extraction code for looping and ImGui replacement.

Malkierian committed:
Commit modules and update refs.
Add extraction, argument, and verification steps.
Remove commented menubar code.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into imgui-extract

Malkierian committed:
Update LUS ref.

Malkierian committed:
Update for `zapd_report` changes in ZAPD branch.

Malkierian committed:
clang

Malkierian committed:
Update ZAPD ref.

Malkierian committed:
Update module refs.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into imgui-extract

Malkierian committed:
Remove OS prompt for cancelling out of file finder dialog on first run and just revert back to PS_FIRST popup.
Re-add ROM search in new flow.
Move port archive checking to ImGui with default fonts.

Malkierian commented:
This is now undrafted, but I'm adding `do not merge` tag because there's one thing that still needs to be resolved with the port archive detection in ResourceManager initialization before it can be merged. Otherwise, this is fully functional.

Malkierian committed:
Convert ROM archive update flow to ImGui.

Malkierian committed:
General cleanup.
Add extra information to extraction modal.
Add PS_FILE_CHECK step.

Malkierian committed:
Allow Exit button to close program during extraction flow.

Malkierian committed:
Merge branch 'develop' of github.com:Malkierian/Shipwright into imgui-extract

Malkierian committed:
Add parameter default of `isJapaneseFont = false` to `CreateFontWithSize()`.

Malkierian committed:
Re-arrange initializers to allow menu to be "setup" before everything is available.

Malkierian committed:
Make modal centering reactive to window size after showing.

Malkierian committed:
clang

Malkierian committed:
Bump LUS.
Further tweak initialization order, and prevent menu from drawing if menu elements aren't initialized.

serprex commented:
On Linux it's hanging on extract of `soh.o2r`, gdb tends to stop in `SohModalWindow::PopupsQueued`

Malkierian commented:
There is no extraction of soh.o2r. It's supposed to exit.

serprex commented:
```
Reading archive: .../soh.o2r
Adding Archive .../soh.o2r to Archive Manager
```
& then it sits. No window appears

Malkierian commented:
I may need you to debug this on Linux. It's virtually impossible to tell where exactly it's hanging just with that log.

serprex commented on soh/soh/Extractor/Extract.cpp:
```diff
@@ -401,7 +401,7 @@ bool Extractor::ManuallySearchForRom() {
 
     if (!GetRomPathFromBox()) {
-        ShowErrorBox("No rom selected", "No Rom selected. Exiting");
+        // ShowErrorBox("No rom selected", "No Rom selected. Exiting");
```
remove commented code (& elsewhere)

serprex commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -263,25 +268,525 @@ const char* constCameraStrings[] = {
+#elif (defined(__WIIU__) || defined(__SWITCH__))
+                    ExtractSteps extractStep = ES_VERIFY;
+#else
+                    ExtractSteps extractStep = ES_EXTRACT;
```
```suggestion
                    extractStep = ES_VERIFY;
#else
                    extractStep = ES_EXTRACT;
```

Malkierian replied:
Ohhhh, that would be why it's hanging.

serprex commented:
After fixing extractStep, when I click `Yes` on `No O2R Files` prompt nothing happens (`No` closes window tho). Looking into..

serprex commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -263,25 +268,525 @@ const char* constCameraStrings[] = {
+                        std::string msg = "Archive for current ROM, " + archive + ", already exists. Extract again?";
+                        SohGui::RegisterPopup("Confirm Re-extract", msg.c_str(), "Yes", "No", [&]() {
+                            extracting = true;
+                            threadPool->submit_task([&]() -> void {
```
need to either wait on these tasks somehow, or the variables they set at end should be atomics

Malkierian replied:
Can't wait because the render portion still has to happen while the extraction is happening.

Malkierian replied:
Why do they need to be atomics? The main thread doesn't change those variables, and no other thread is possible to be made while one extraction is running.

serprex replied:
In C one thread reading a variable (eg, `promptStep`) that's being modified in another thread at the same time is UB

Malkierian replied:
Well, `promptStep` isn't even being read while an extraction thread is active, but I can do it for extracting, `extractCount`, and `totalExtract`.

Malkierian replied:
Shoot, making them atomic requires further changes to ZAPD XD.

serprex replied:
looks like you could read atomic to local, then pass pointer to local to ZAPD? I guess the issue is reading ZAPD updates in meantime

I've seen progress callbacks in past so caller can decide to copy progress to atomics or not *(ofc, that'd still be further changes to ZAPD here)*

serprex replied:
https://en.cppreference.com/w/cpp/thread/future/wait_for

can pass a zero duration to poll task status

thinking that'd also be where to check on exceptions:

> Every task submitted to the queue using [submit_task()](https://github.com/bshoshany/thread-pool#submitting-tasks-to-the-queue) automatically generates an std::future, which can be used to wait for the task to finish executing, obtain its eventual return value, and/or catch any thrown exceptions.

from https://github.com/bshoshany/thread-pool

Malkierian replied:
I opted just to pass pointers of the atomic variables where needed. The functionality ended up being exactly the same otherwise.

Malkierian commented:
> After fixing extractStep, when I click `Yes` on `No O2R Files` prompt nothing happens (`No` closes window tho). Looking into..

Does this occur with ROMs in the folder, without them there, or both?

Malkierian commented:
Nevermind, I think I found it.

Malkierian committed:
Fix hangs, remove commented code.

Malkierian commented:
Oh, right, since this also holds a LUS bump, I *think* #6121 is unnecessary (since that's only a LUS bump).

serprex commented:
> Nevermind, I think I found it.

confirmed file dialog opens, stuck in `Starting Up` after

Malkierian commented:
> > Nevermind, I think I found it.
> 
> confirmed file dialog opens, stuck in `Starting Up` after

Are you sure? It takes like 15 seconds to start the progress bar even on release. Probably takes longer in debug.

serprex commented:
had left it for about half an hour, & I'm on a pretty beefy machine

Malkierian commented:
Well, I don't know what's going on with that, then, as I just tested it locally and it was fine. Probably something in the thread dispatching or the setup for ZAPD.

serprex commented:
ZAPD is throwing `Error: Unable to read config file.` & code doesn't pick up on task failing atm

caused because `assets` directory in my build dir exists but is empty (I created it earlier)

Malkierian commented:
Well then fill it, lol. That's supposed to be setup with the setup command, isn't it?

serprex commented:
I created it when I got error message `Missing 'assets/' folder needed to generate OTR file`

Malkierian commented:
My point is that the assets folder is supposed to be created and filled in setup or build or something. That has nothing to do with the functionality of the code at this point.

serprex commented:
extraction working now, needed to run `soh.elf` from `build/soh` instead of running `soh/soh.elf` from `build` directory

serprex commented:
> My point is that the assets folder is supposed to be created and filled in setup or build or something. That has nothing to do with the functionality of the code at this point.

Sure, but better error messages or having error handling show error messages when human involved is being stupid would be good

Malkierian committed:
Make progress variables atomic for thread safety.

serprex reviewed (approved):
works for me, could use exception handling

Malkierian committed:
clang

Malkierian committed:
Swap AppDirectory for AppBundle for `installPath` in ZAPD calls.

Malkierian committed:
clang

Malkierian committed:
Remove border from progress bar.
Extract filename from file path for display in extraction progress.
Fix searchPath not being set in PS_LOCAL.
Improve newlining for text to better fit default window.
